tests/python/test_dmatrix.py

```diff
@@ -12,6 +13,29 @@
 
 
 class TestDMatrix(unittest.TestCase):
+    def test_warn_missing(self):
+        from xgboost import data
+        with pytest.warns(UserWarning):
+            data._warn_unused_missing('uri', 4)
+
+        with pytest.warns(None) as record:
+            data._warn_unused_missing('uri', None)
+            data._warn_unused_missing('uri', np.nan)
+
+        assert len(record) == 0
```

Comment:
Which `record` is this referring to?

Reply:
@hcho3 At line 21.

Reply:
Is it legal to refer to a `with` variable past its scope? It's as if:
```
with open(..) as f:
   ...
print('foobar', file=f)
```
which seems undefined.

Reply:
Let me move that inside.

Reply:
Done.